### PR TITLE
Add hints for type errors where the user might have meant `{=}` instead of `{}`

### PR DESCRIPTION
**Pull request by timbertson**
I thought https://github.com/dhall-lang/dhall-lang/issues/1024 would make for a good first use case of context-specific hints next to errors, much like `rustc`. The idea is that if you have a type error involving `{}`, and it's a context where `{=}` could work, we include a hint in the error output:

```
$ echo '{ x = 1 } // {}' | ./dist/build/Dhall/dhall

Use "dhall --explain" for detailed errors

Error: You can only override records

Hint: {} is the empty record type, use {=} for the empty record value

1│ { x = 1 } // {}

(input):1:1
```

This hint appears for the following type errors:

 - `({Type = { x: Natural }, default={ x = 1 }})::{}`
 - `{} // { x = 1 }`
 - `{} /\ { x = 1 }`
 - `(\(x: Natural) -> x + 1) {}`

The hint is only shown when the erroneous value in question is literally `{}` so it should have a fairly high relevance. In the final example `{=}` would _also_ be the wrong type though, so it's not 100%.

The only issue I had is when `{}` is on the right hand side of a combine operation, i.e. `{ x = 1 } /\ {}`

The `--explain` text for this one actually looks incorrect to me, it ends in:

```
You supplied this expression as one of the arguments:

↳ { x = 1 }

... which is not a record, but is actually a:

↳ Type
```

My hint shows up when I reverse the operands, so it seems like maybe the code constructing this error is always passing in the left hand side, rather than the expression matching the non-record type.

### CODE REVIEW

**Comment by sjakobi** [comment]
@timbertson Thanks for the report of the incorrect type error explanation! I have created a fix in https://github.com/dhall-lang/dhall-haskell/pull/2244.

File: dhall/src/Dhall/TypeCheck.hs [hunk 1]
```diff
@@ -2608,9 +2659,11 @@ prettyTypeMessage (CompletionSchemaMustBeARecord expr0 expr1) = ErrorMessages {.
 
 prettyTypeMessage (InvalidRecordCompletion fieldName expr0) = ErrorMessages {..}
  where
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs:2661]: Alignment wibble:

```suggestion
  where
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 2]
```diff
@@ -2584,9 +2633,11 @@ prettyTypeMessage (ListAppendMismatch expr0 expr1) = ErrorMessages {..}
 
 prettyTypeMessage (CompletionSchemaMustBeARecord expr0 expr1) = ErrorMessages {..}
  where
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs:2635]:
```suggestion
  where
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 3]
```diff
@@ -1380,15 +1381,19 @@ data TypeMessage s a
     | CantMultiply (Expr s a) (Expr s a)
     deriving (Show)
 
+formatHints :: [Doc Ann] -> Doc Ann
+formatHints hints = vsep (map format hints) where
+    format hint = "\n\n\ESC[1;33mHint\ESC[0m: " <> hint
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs:1386]: Minor style issue:

```suggestion
formatHints hints = vsep (map format hints)
  where
    format hint = "\n\n\ESC[1;33mHint\ESC[0m: " <> hint
```

File: dhall/src/Dhall/TypeCheck.hs [hunk 4]
```diff
@@ -1,4 +1,5 @@
 {-# LANGUAGE DeriveDataTypeable  #-}
+{-# LANGUAGE NamedFieldPuns      #-}
```
Comment (sjakobi) [inline dhall/src/Dhall/TypeCheck.hs:2]: What do we need `NamedFieldPuns` for?

**Review by sjakobi — COMMENTED** [review]
I like it! :)

File: dhall/src/Dhall/TypeCheck.hs (hunk 4 above)
Comment (timbertson) [reply dhall/src/Dhall/TypeCheck.hs]: Probably some work-in-progress code, I've removed this now.

**Review by sjakobi — APPROVED** [review]

**Review by Gabriel439 — APPROVED** [review]
Thank you for doing this! 🙂

**Pull request merged**